openlibrary/plugins/openlibrary/partials.py

```diff
@@ -0,0 +1,203 @@
+    if 'env' not in web.ctx:
+        delegate.fakeload()
+    work = web.ctx.site.get('/works/%s' % workid) or {}
+    component = render_template('books/RelatedWorksCarousel', work)
+    return {0: str(component)}
+
+
+def get_cached_relatedcarousels_component(*args, **kwargs):
+    memoized_get_component_metadata = cache.memcache_memoize(
+        _get_relatedcarousels_component,
+        "book.bookspage.component.relatedcarousels",
+        timeout=dateutil.HALF_DAY_SECS,
+    )
+    return (
+        memoized_get_component_metadata(*args, **kwargs)
+        or memoized_get_component_metadata.update(*args, **kwargs)[0]
+    )
+
+
+class CarouselCardPartial:
+    MAX_VISIBLE_CARDS = 5
+
+    def __init__(self):
+        self.i = web.input(params=None)
+
+    def generate(self) -> dict:
+        # Determine query type
+        params = self.i or {}
+        query_type = params.get("queryType", "")
+
+        # Do search
+        search_results = self._make_book_query(query_type, params)
+
+        # Render cards
+        cards = []
+        layout = params.get("layout")
+        key = params.get("key") or ""
+
+        for index, book in enumerate(search_results):
+            lazy = index > self.MAX_VISIBLE_CARDS
+            cards.append(
+                render_template(
+                    "books/custom_carousel_card",
+                    web.storage(book),
+                    lazy,
+                    layout,
+                    key=key,
+                )
+            )
+
+        # Return partials dict:
+        return {"partials": [str(template) for template in cards]}
+
+    def _make_book_query(self, query_type: str, params: dict) -> list:
+        if query_type == "SEARCH":
+            return self._do_search_query(params)
+        if query_type == "BROWSE":
+            return self._do_browse_query(params)
+        if query_type == "TRENDING":
+            return self._do_trends_query(params)
+        if query_type == "SUBJECTS":
+            return self._do_subjects_query(params)
+
+        raise ValueError("Unknown query type")
+
+    def _do_search_query(self, params: dict) -> list:
+        fields = 'key,title,subtitle,author_name,cover_i,ia,availability,id_project_gutenberg,id_project_runeberg,id_librivox,id_standard_ebooks,id_openstax'
+        query = params.get("q", "")
+        sort = params.get("sorts", "new")  # XXX : check "new" assumption
+        limit = int(params.get("limit", 20))
+        page = int(params.get("page", 1))
+        query_params = {
+            "q": query,
+            "fields": fields,
+        }
+        if fulltext := params.get("hasFulltextOnly"):
+            query_params['has_fulltext'] = 'true'
+
+        results = work_search(
+            query_params, sort=sort, limit=limit, facet=False, offset=page
+        )
+        return results.get("docs", [])
+
+    def _do_browse_query(self, params: dict) -> list:
+        url = params.get("q", "")
+
+        results = get_available(url=url)
```

Comment:
Womp womp, this doesn't take page or offset or anything any that is almost certainly the problem 🎉